src/main/java/graphql/execution/MergedField.java

```diff
@@ -171,68 +322,114 @@ public MergedField transform(Consumer<Builder> builderConsumer) {
         return builder.build();
     }
 
+    /**
+     * Runs a consumer for each field
+     *
+     * @param fieldConsumer the consumer to run
+     */
+    public void forEach(Consumer<Field> fieldConsumer) {
+        fieldConsumer.accept(singleField);
+    }
+
     public static class Builder {
 
-        private final ImmutableList.Builder<Field> fields = new ImmutableList.Builder<>();
-        private final ImmutableList.Builder<DeferredExecution> deferredExecutions = new ImmutableList.Builder<>();
+        /*
+            The builder logic is complicated by these dual singleton / list duality code,
+            but it prevents memory allocation and every bit counts
+            when the CPU is running hot and an operation has lots of fields!
+         */
+        private ImmutableList.@Nullable Builder<Field> fields;
+        private @Nullable Field singleField;
+        private ImmutableList.@Nullable Builder<DeferredExecution> deferredExecutions;
 
         private Builder() {
         }
 
         private Builder(MergedField existing) {
```

Comment:
Technically we dont have to change the builder since the field collector now uses a direct instationation methods to make a MergedField without a builder.  But I thought why not - make it as memory efficient as possible even if its more complicated with its lazy instatation